src/cargo/core/compiler/fingerprint.rs

```diff
@@ -277,6 +376,24 @@ pub fn prepare_target<'a, 'cfg>(
         return Ok(Job::new(Work::noop(), Fresh));
     }
 
+    // Clear out the old fingerprint file if it exists. This protects when
+    // compilation is interrupted leaving a corrupt file. For example, a
+    // project with a lib.rs and integration test:
+    //
+    // 1. Build the integration test.
+    // 2. Make a change to lib.rs.
+    // 3. Build the integration test, hit Ctrl-C while linking (with gcc).
+    // 4. Build the integration test again.
+    //
+    // Without this line, then step 4 will think the integration test is
+    // "fresh" because the mtime of the output file is newer than all of its
+    // dependencies. But the executable is corrupt and needs to be rebuilt.
+    // Clearing the fingerprint ensures that Cargo never mistakes it as
+    // up-to-date until after a successful build.
+    if loc.exists() {
+        paths::write(&loc, b"")?;
+    }
```

Comment:
Trying to make sure I fully understand this. So today we consider step 4 fresh, but it should actually rerun because the linker didn't actually finish successfully. When linking is interrupted the compiler has already emitted the dep-info file, so when we finish the unit and wrap up the fingerprint stuff it looks like it succeeded because the output file exists and the dep-info exists? Is that right?

This feels like it may be best handled by we truncate the file if the process failed for whatever reason. I'm a little hazy on this though. Is the idea that this truncation happens in step 3, which when interrupted, means step 4 picks up the truncated file?

Reply:
>  and the dep-info exists? Is that right?

I don't think it matters whether or not rustc emitted the `.d` file.  Cargo doesn't translate it into the fingerprint directory unless the compilation finished successfully.  

The issue is that there is a valid fingerprint from the *previous* run (step 1). During step 3, the new unit test executable is created but not completed.  In step 4, all components of the unit test fingerprint are fresh from step 1 (nothing changed), and the mtime comparison of the incomplete executable is newer than the library's rlib (and all the unit test source files), so it thinks it is fresh.  

> truncate the file if the process failed

I don't think we can only truncate on failure, because Cargo is forcefully terminated. (I guess we could catch signals but that doesn't help with SIGKILL, and I'd prefer not to deal with signals.)

>  Is the idea that this truncation happens in step 3, which when interrupted, means step 4 picks up the truncated file?

Yes, step 4 compares an empty string with the expected hash string, and determines it is dirty because they are not equal.

Reply:
Man see this is when I have mega-doubts about everything. Doing anything with mtimes is so tricky that it seems like a folly almost. In any case agreed we shouldn't catch signals (was misinterpreting this for a second as failed rustc vs interrupted cargo).

I think the way we should think about this is that when we start a compilation we *delete* the expected outputs. That way if something bad happens, the next time we pick up the outputs won't exist and we'll see we need to recompile. This only works if the outputs are Cargo-created, not well-known files (like the linked artifact). The dep-info here, consequently, is a Cargo-output which only Cargo creates, so it makes sense to delete it before a build and create it after a successful build.

Would it perhaps make more sense to delete the file here rather than truncating it to zero bytes?

Reply:
The reason it truncates instead of deleting is that the fingerprint debug logging doesn't work in the scenario:

1. Successful build.
2. Change something.
3. Build with a compile error.  Logs fingerprint reason.  **Delete** fingerprint.
4. Build again.  Fingerprint log just says "file not found" with no other detail.

I wanted to make sure step 4 continued to log the real reason for the fingerprint error (like "current filesystem status shows we're outdated") rather than "file not found".  

I don't think it is too important (the first build failure logs the real reason), so I can switch it to delete if you prefer.  I'm curious what concerns you have about truncating, though?

Reply:
Nah that makes sense to me, so I think truncation here is fine. Could you update the comment with some of this discussion though? (about what files are flowing into what steps and why we're truncating instead of deleting).

I'm slightly confused again though. I would have thought that step 4 can't read the fingerprint file of step 1 (without this patch) because the mtime of the input files (from step 2) shows that the fingerprint file is stale. The mtime of the file from step 2 would be later than that of the fingerprint file from step 1, right?

(or do we write out the fingerprint file again in step 3?)

Reply:
Checking mtime is done at the same time as checking the fingerprint hash [here](https://github.com/rust-lang/cargo/blame/3a976c17670a5bede70049a1ec7c11351bfaa49c/src/cargo/core/compiler/fingerprint.rs#L1408).  So in a sense it checks *both* the fingerprint hash and the fs_status at the same time.  If either of them fail, it falls through to the code below to log the failure.  

If it deleted the file, the first line in `compare_old_fingerprint` would fail, circumventing the logging.

(Fingerprints are only written after a successful build.)

Reply:
Er sorry I was roundaboutedly asking a question on [your response above](https://github.com/rust-lang/cargo/pull/8087#discussion_r406297330) where you said

> The issue is that there is a valid fingerprint from the previous run (step 1). During step 3, the new unit test executable is created but not completed. In step 4, all components of the unit test fingerprint are fresh from step 1 (nothing changed), and the mtime comparison of the incomplete executable is newer than the library's rlib (and all the unit test source files), so it thinks it is fresh.

You say "all components of the unit test fingerprint are fresh from step 1" but I thought that the final fingerprint file that Cargo writes for the executable woul dnot be fresh, right? The executable is there (and corrupt) but a source file is still newer than the final fingerprint file because we don't write that until rustc finishes I thought?

Reply:
Sorry, I made a typo in my reply.  Where I said "unit test" I meant "integration test".  Going back to the original example, "lib.rs" is changed, and compiles successfully.  Then Cargo goes to build the *integration test* executable, and is interrupted.  The next time around (step 4) "lib.rs" is up-to-date (and correct on-disk).  And none of the integration test sources have changed, so that check appears up-to-date as well.

That was a bit of a subtle point, I'll emphasize that in the comment.

Reply:
I pushed a commit with extra details.  Hopefully it's not too confusing, I tried to be more precise with the wording.